caas/kubernetes/provider/k8stypes_test.go

```diff
@@ -422,3 +509,118 @@ containers:
 	err = spec.Validate()
 	c.Assert(err, gc.ErrorMatches, `mount path is missing for file set "configuration"`)
 }
+
+func (s *ContainersSuite) TestParsePodSpecFailedBothServiceAccountAndServiceAccountNameProvided(c *gc.C) {
+	specStr := `
+serviceAccountName: an-existing-svc-account
+serviceAccount:
+  name: build-robot
+  automountServiceAccountToken: true
+  capabilities:
+    roleBinding:
+      name: read-pods
+      type: ClusterRoleBinding
+    role:
+      name: pod-reader
+      type: ClusterRole
+      rules:
+      - apiGroups: [""]
+        resources: ["pods"]
+        verbs: ["get", "watch", "list"]
+`[1:]
+	_, err := provider.ParseK8sPodSpec(specStr)
+	c.Assert(err, gc.ErrorMatches, "either use ServiceAccountName to reference to existing service account or define ServiceAccount spec to create a new one")
+
+}
+
+func (s *ContainersSuite) TestValidateServiceAccountFailed(c *gc.C) {
+	containerSpec := `
+containers:
+  - name: gitlab-helper
+    image: gitlab-helper/latest
+    ports:
+    - containerPort: 8080
+      protocol: TCP
+`[1:]
+
+	for i, tc := range []struct {
```

Comment:
This is horrible, move it to a variable please.

Reply:
moved out from the func